pkg/virt-controller/watch/pool/pool.go

```diff
@@ -1650,3 +1657,40 @@ func (c *Controller) handlePoolDeletion(pool *poolv1.VirtualMachinePool, vms []*
 
 	return nil
 }
+
+func isAutohealingEnabled(pool *poolv1.VirtualMachinePool) bool {
+	return pool.Spec.Autohealing != nil && *pool.Spec.Autohealing
+}
+
+func (c *Controller) autoHealFailingVMs(pool *poolv1.VirtualMachinePool, vms []*virtv1.VirtualMachine) error {
+	vmsToCleanup := filterVMs(vms, func(vm *virtv1.VirtualMachine) bool {
+		return vm.Status.StartFailure != nil && vm.Status.StartFailure.ConsecutiveFailCount >= defaultStartUpFailureThreshold
+	})
+
+	if len(vmsToCleanup) > 0 {
+		basePolicy := resolveBasePolicy(pool.Spec.ScaleInStrategy)
+		sortVMsForDownscale(vmsToCleanup, basePolicy)
+
+		for _, vm := range vmsToCleanup {
+			if err := c.cleanupFailingVM(vm); err != nil {
+				log.Log.Object(vm).Errorf("Failed to cleanUp VM: %v", err)
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) cleanupFailingVM(vm *virtv1.VirtualMachine) error {
+	if err := c.removeFinalizer(vm); err != nil {
+		log.Log.Object(vm).Errorf("Failed to remove finalizer: %v", err)
+		return err
+	}
+
+	if err := c.clientset.VirtualMachine(vm.Namespace).Delete(context.Background(), vm.Name, metav1.DeleteOptions{PropagationPolicy: pointer.P(metav1.DeletePropagationForeground)}); err != nil {
+		log.Log.Object(vm).Errorf("Failed to delete VM: %v", err)
+		return err
+	}
+	return nil
+}
```

Comment:
Isn't this a duplication of `scaleIn`'s cleanup code but missing the delete event?